vllm/v1/core/scheduler.py

```diff
@@ -542,6 +560,26 @@ def update_from_output(
             new_logprobs = None
             new_token_ids: List[int] = []
 
+            # Handle guided decoding FSM advancement if applicable
+            if (request.use_guided_decoding and request.grammar
+                    and request.is_grammar_ready
+                    and not request.grammar.prefilled):
+                index = model_runner_output.req_id_to_index.get(req_id)
+                if index is not None:
+                    # TODO - fix spec decode + structured output compatibility
```

Comment:
I left this how it is to make the changes in the merge with main as small as possible. I didn't want to try to fix spec decode compatibility there, but instead do it as another commit, so this is just handling a single token like it was pre-merge at the moment.